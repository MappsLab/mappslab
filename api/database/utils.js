// @flow

import * as R from 'ramda'
import type { Filter, PaginationArgs } from 'Types/sharedTypes'

export const createVariables = R.pipe(
	R.toPairs,
	R.reduce(
		(acc, [key, value]) =>
			// if the value is undefined or null,
			value === undefined || value === null
				? // return the accumulator
				  acc
				: // Else, return the value cast to a string
				  R.assoc(`$${key}`, value.toString())(acc),
		{},
	),
)

export const defaultPaginationArgs = {
	first: 50,
	after: '0x0',
}

export const makePaginationArgs = (args: PaginationArgs): PaginationArgs => {
	// Adds +1 to the 'first' argument, we need this to see if there is a next page.
	const { first, ...rest } = { ...defaultPaginationArgs, ...args }
	return {
		first: first + 1,
		...rest,
	}
}

export const createFilterString = (filter: Array<Filter>, connect: string = 'AND'): string => {
	const funcs = filter.map(({ key, value, operator = 'eq' }) => `${operator}(${key}, "${value}")`)
	return `@filter(${funcs.join(connect)})`
}

export const validateUid = (uid: string): boolean => /^0x[\d\w]+$/.test(uid)

export const validateUidOrWildcard = (uid: string): boolean => uid === '*' || validateUid(uid)
