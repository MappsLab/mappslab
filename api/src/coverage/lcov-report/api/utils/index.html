<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for api/utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> api/utils
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">84.21% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>48/57</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">84.85% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>28/33</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.82% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>18/22</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">85.11% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>40/47</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="auth.js"><a href="auth.js.html">auth.js</a></td>
	<td data-value="85" class="pic high"><div class="chart"><div class="cover-fill" style="width: 85%;"></div><div class="cover-empty" style="width:15%;"></div></div></td>
	<td data-value="85" class="pct high">85%</td>
	<td data-value="20" class="abs high">17/20</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="84.21" class="pct high">84.21%</td>
	<td data-value="19" class="abs high">16/19</td>
	</tr>

<tr>
	<td class="file high" data-value="graphql.js"><a href="graphql.js.html">graphql.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="13" class="abs high">12/13</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file high" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="81.25" class="pic high"><div class="chart"><div class="cover-fill" style="width: 81%;"></div><div class="cover-empty" style="width:19%;"></div></div></td>
	<td data-value="81.25" class="pct high">81.25%</td>
	<td data-value="16" class="abs high">13/16</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="70" class="pct medium">70%</td>
	<td data-value="10" class="abs medium">7/10</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	</tr>

<tr>
	<td class="file medium" data-value="parsing.js"><a href="parsing.js.html">parsing.js</a></td>
	<td data-value="72.73" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 72%;"></div><div class="cover-empty" style="width:28%;"></div></div></td>
	<td data-value="72.73" class="pct medium">72.73%</td>
	<td data-value="11" class="abs medium">8/11</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="8" class="abs medium">6/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="72.73" class="pct medium">72.73%</td>
	<td data-value="11" class="abs medium">8/11</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Nov 12 2018 20:15:34 GMT-0800 (Pacific Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
