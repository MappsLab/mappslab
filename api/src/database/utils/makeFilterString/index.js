// @flow
import type { Filter } from 'Types/sharedTypes'
import createRelationshipFilter from './relationships'
import createOperatorFilter from './operators'
import type { FilterStrings } from '../dbUtils'

export type Range = {
	start: Date,
	end: Date,
}

export type FilterObject = {
	[key: string]: Filter,
}

const joinStrings = (connect: string) => (a: string, b: string): string =>
	`${a}${a.length && b.length ? connect : ''}${b}`

const andJoin = joinStrings(' AND ')
const nJoin = joinStrings('\n')

const groupFilterStrings = ({ filterString, varBlocks }: FilterStrings) => ({
	varBlocks,
	filterString: `(${filterString})`,
})

const joinFilterStrings = (filters: Array<FilterStrings>) => {
	const joined = filters.reduce(
		(acc, current) => ({
			filterString: andJoin(acc.filterString, current.filterString),
			varBlocks: nJoin(acc.varBlocks, current.varBlocks),
		}),
		{
			filterString: '',
			varBlocks: '',
		},
	)
	return groupFilterStrings(joined)
}

/**
 * makeFilterPartial
 * returns joined filterStrings and varBlocks for a single field
 */
const makeFilterPartial = (field: string) => ([operator, value]: [
	string,
	string,
]): FilterStrings => {
	// const r = createOperatorFilter
	const filters =
		createRelationshipFilter(operator, field, value) ||
		createOperatorFilter(operator, field, value)
	return filters
}

const makeFieldFilter = ([field, filters]: [string, Filter]) => {
	const filterStrings = Object.entries(filters).map(makeFilterPartial(field))
	return joinFilterStrings(filterStrings)
}

const makeFilterString = (where?: FilterObject): FilterStrings => {
	if (!where) return { varBlocks: '', filterString: '' }
	// Separate out the 'where' object into field -> filter pairs.
	// Each of these fields will have its filters applied as a group.
	const fieldFilters = Object.entries(where).map(makeFieldFilter)
	const { filterString, varBlocks } = joinFilterStrings(fieldFilters)
	return {
		varBlocks,
		filterString: `@filter${filterString}`,
	}
}

export default makeFilterString
