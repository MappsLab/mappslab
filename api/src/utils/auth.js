// @flow
import crypto from 'crypto-promise'
import jwt from 'jsonwebtoken'
import type { JWT, UserType } from 'Types/UserTypes'
import config from '../config'

const JWT_KEY = config.get('jwtKey')

export const createToken = (): Promise<string> =>
	new Promise((resolve, reject) => {
		crypto
			.randomBytes(48)
			.then((buffer) => {
				const token = buffer.toString('hex')
				resolve(token)
			})
			.catch((cryptErr) => {
				reject(cryptErr)
			})
	})

export const createJWT = (user: UserType): JWT => {
	const expiresIn = 1 * 24 * 60 * 60 // 1 day
	const { name, uid, roles } = user
	if (!name || !uid || !roles || !roles.length)
		throw new Error(`createJWT requires a name, uid, and roles`)
	const token = jwt.sign({ name, uid, roles }, JWT_KEY || 'abc', { expiresIn })
	const expires = new Date(Date.now() + expiresIn * 1000).toISOString() // send an actual date
	return {
		token: `Bearer ${token}`,
		expires,
	}
}

export const verifyJWT = (token: string): Promise<UserType> =>
	new Promise((resolve, reject) => {
		jwt.verify(token, JWT_KEY, (err, decoded) => {
			if (err) {
				reject(err)
				return
			}
			resolve(decoded)
		})
	})
