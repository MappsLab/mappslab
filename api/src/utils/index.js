// @flow
import { pipeP, flatten, filter, assoc, prop } from 'ramda'
import type { NestedArray } from 'ramda'

export const filterNullAndUndefined = (obj: Object): Promise<Object> =>
	new Promise((resolve) =>
		resolve(filter((a) => a !== null && a !== undefined)(obj)),
	)

// Makes regular functions async for promisePipe
export const promisify = (f: Function): Function => async (...args) =>
	f(...args)

export const promisePipe = (...funcs: Array<Function>) =>
	pipeP(...funcs.map(promisify))

export const arrayify = (...things: NestedArray<mixed>): Array<mixed> =>
	flatten(things)

export const slugify = (text: string) =>
	text
		.toString()
		.toLowerCase()
		.replace(/\s+/g, '-')
		.replace(/[^\w-]+/g, '')
		.replace(/--+/g, '-')
		.replace(/^-+/, '')
		.replace(/-+$/, '')

export const createSlugFrom = (key: string) => (obj: Object) =>
	assoc('slug', slugify(prop(key, obj)))(obj)
