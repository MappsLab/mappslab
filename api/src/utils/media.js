// @flow
import sharp from 'sharp'
import toArray from 'stream-to-array'
import type { ImageUpload } from 'Types/ImageTypes'

type ParsedImage = {
	data: Buffer,
	info: {
		width: number,
		height: number,
		format: string,
	},
}

export const parseImage = async (source: Buffer): Promise<ParsedImage> => sharp(source).toBuffer({ resolveWithObject: true })

export const resizeImage = async (source: Buffer, width: number, height?: number): Promise<ParsedImage> =>
	sharp(source)
		.resize(width, height)
		.toBuffer({ resolveWithObject: true })

export const streamToBuffer = async (source: ImageUpload): Promise<Buffer> =>
	new Promise(async (resolve) => {
		const { createReadStream } = await source
		const stream = createReadStream()
		toArray(stream).then((parts) => {
			const buffers = parts.map((part) => (Buffer.isBuffer(part) ? part : Buffer.from(part)))
			resolve(Buffer.concat(buffers))
		})
	})
