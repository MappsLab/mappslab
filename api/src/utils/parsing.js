// @flow
import { mapObjIndexed, head } from 'ramda'

const debug = require('debug')('api')

export const parseSingularFields = (singularFields: Array<string>) =>
	mapObjIndexed((value, key) => {
		if (singularFields.includes(key) && Array.isArray(value)) {
			if (!Array.isArray(value)) {
				debug(`Field '${key}' is not an array`)
				return value
			}
			if (value.length > 1)
				debug(`Field '${key}' has multiple entries but should be one. Your database might have unused entries.`)
			return head(value)
		}
		return value
	})
