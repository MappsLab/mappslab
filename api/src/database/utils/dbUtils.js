// @flow

import * as R from 'ramda'
import type { Facet } from 'Types/database'
import type { PaginationFilterArgs } from 'Types/sharedTypes'
import makeFilterString from './makeFilterString'

/**
 * Format facets { key: 'value' } => (key=value)
 */

const isValidDate = (date): boolean =>
	typeof date === 'object' && Object.prototype.toString.call(date) === '[object Date]' && !Number.isNaN(date)

const toString = (val: mixed) => {
	switch (typeof val) {
		case 'string':
			return val
		case 'boolean':
		case 'number':
			return val.toString()
		case 'object':
			// $FlowFixMe
			if (isValidDate(val)) return val.toISOString()
			throw new Error('Facet value must be a string, boolean, number, or Date')
		default:
			throw new Error('Facet value must be a string, boolean, number, or Date')
	}
}

export const serializeFacets = (facets: Array<Facet> = []): string =>
	Object.entries(facets).length
		? Object.entries(facets)
				.map(([key, value]) => `${key}=${toString(value)}`)
				.join(', ')
				.replace(/(.*)/, '($1)')
		: ''

/**
 * Properly formats incoming variables with '$' for dgraph
 */

// $FlowFixMe
export const createVariables = R.pipe(
	R.toPairs,
	R.reduce(
		(acc, [key, value]) =>
			// if the value is undefined or null,
			value === undefined || value === null
				? // return the accumulator
				  acc
				: // Else, return the value cast to a string
				  R.assoc(`$${key}`, value.toString())(acc),
		{},
	),
)

/**
 * Create a paginatino string to append to a dGraph func. Be sure to include the
 * initial comma that appears after the root function, i.e. `, first: 5, after: 0x123`
 * @param {number} first
 * @param {string} after
 */
export const DEFAULT_PAGE_COUNT = 50

const makePaginationString = (first?: number = DEFAULT_PAGE_COUNT, after?: string): string => {
	return [
		'', // include a blank string so we start it with a comma
		first // if a 'first' value is present, include it +1 so we can tell if there is a 'nextPage'
			? `first: ${first + 1}`
			: null,
		after //
			? `after: ${after}`
			: null,
	]
		.filter((i) => i !== null)
		.join(', ')
}

/**
 * Create a @filter string to append to a dGraph query. Be sure to include the
 * initial comma that appears after the root function, i.e. `, first: 5, after: 0x123`
 * @param {number} first
 * @param {string} after
 */

export type FilterStrings = {
	filterString: string,
	varBlocks: string,
}

type QueryStrings = FilterStrings & {
	paginationString: string,
}

export const createQueryStrings = (args: PaginationFilterArgs, filterDeleted?: boolean = true): QueryStrings => {
	if (!args)
		return {
			paginationString: '',
			filterString: '@filter((eq(deleted, false)))',
			varBlocks: '',
		}
	const { first, after, where: baseWhere } = args
	const where = filterDeleted
		? {
				deleted: { eq: false },
				...baseWhere,
		  }
		: baseWhere

	return {
		paginationString: makePaginationString(first, after),
		...makeFilterString(where),
	}
}

/**
 * Validate that the string is a dGraph uid
 */

export const validateUid = (uid: string): boolean => /^0x[\d\w]+$/.test(uid)

export const validateUidOrWildcard = (uid: string): boolean => uid === '*' || validateUid(uid)
