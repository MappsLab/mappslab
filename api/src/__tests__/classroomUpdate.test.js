import { request } from './utils/db'
import { getDBUsers } from './utils/user'
import { getDBClassrooms } from './utils/classroom'

let classroom
let teacher
let otherTeacher
let student

const query = /* GraphQL */ `
	query classroom($uid: String!) {
		classroom(input: { uid: $uid }) {
			uid
			title
			description
			teachers {
				edges {
					cursor
					node {
						uid
						name
						roles
					}
				}
			}
		}
	}
`

beforeAll(async (done) => {
	const firstClassrooms = await getDBClassrooms()
	const result = await request(query, {
		variables: { uid: firstClassrooms[0].uid },
	})
	// eslint-disable-next-line
	classroom = result.data.classroom
	const users = await getDBUsers()
	const classroomTeachers = classroom.teachers.edges.map((edge) => edge.node)
	// eslint-disable-next-line
	teacher = classroomTeachers[0]
	otherTeacher = users.find(
		(u) =>
			u.roles.includes('teacher') &&
			!classroomTeachers.find((ct) => ct.uid === u.uid),
	)
	student = users.find(
		(u) => !u.roles.includes('teacher') && u.roles.includes('student'),
	)
	done()
})

const mutation = /* GraphQL */ `
	mutation UpdateClassroom(
		$uid: String!
		$title: String
		$description: String
	) {
		updateClassroom(
			input: { uid: $uid, title: $title, description: $description }
		) {
			uid
			title
			description
		}
	}
`

const getVariables = () => ({
	uid: classroom.uid,
	description: 'a new description',
})

describe('[updateClassroom]', () => {
	it('should return an error if there is no current viewer', async () => {
		const variables = getVariables()
		const result = await request(mutation, { variables })
		expect(result.errors).toMatchSnapshot()
	})

	it('should return an error if the user is not a teacher', async () => {
		const variables = getVariables()
		const context = { viewer: student }
		const result = await request(mutation, { variables, context })
		expect(result.errors).toMatchSnapshot()
	})

	it('should return an error if the teacher does not own the classroom', async () => {
		const variables = getVariables()
		const context = { viewer: otherTeacher }
		const result = await request(mutation, { variables, context })
		expect(result.errors).toMatchSnapshot()
	})

	it('should properly update a classroom', async () => {
		const original = await request(query, { variables: { uid: classroom.uid } })
		const originalDescription = original.data.classroom.description
		const originalTitle = original.data.classroom.title
		const variables = getVariables()
		const context = { viewer: teacher }
		const result = await request(mutation, { variables, context })
		expect(result.data.updateClassroom.description).toBe(variables.description)
		expect(result.data.updateClassroom.title).toBe(originalTitle)
		const restore = await request(mutation, {
			variables: { ...variables, description: originalDescription },
			context,
		})
		expect(restore.data.updateClassroom.description).toBe(originalDescription)
		expect(result.data.updateClassroom.title).toBe(originalTitle)
	})
})
