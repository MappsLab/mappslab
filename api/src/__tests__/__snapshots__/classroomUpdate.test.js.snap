// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[updateClassroom] should return an error if the teacher does not own the classroom 1`] = `
Array [
  [GraphQLError: You can only modify classrooms that you teach in.],
]
`;

exports[`[updateClassroom] should return an error if the user is not a teacher 1`] = `
Array [
  [GraphQLError: You must be a teacher to create a classroom],
]
`;

exports[`[updateClassroom] should return an error if there is no current viewer 1`] = `
Array [
  [GraphQLError: You must be logged in to create a new classroom. Please log in and try again.],
]
`;
